usb: Enable USB by the application.

eobalski opened the pull request:
The only reasonable solution right now to address #27071 is forcing the application to manually enable USB subsystem.
This patch updates remaining samples (not included in #20375) to call `usb_enable()` in `main`.

@markus-becker-tridonic-com please review the sample You added recently (last commit). I don't know what was expected to be on usb-cdc but right now I can see some garbage output (with master as well).
Furthermore:
- the sample has several Kconfig warnings that could be addressed (both overlays & without any).
- The sample builds with warnings when `overlay-tri-n4m-br.conf` applied (`-Wunused-varaible`).

Fixes: #27071

Commit:
samples: usb: Make USB console samples call usb_enable()

usb_enable() must be called once by the application.
The application may want to register usb_dc_status_callback
and trace usb status codes (usb_dc_status_code).

After this patch all pre APPLICATION messages will be dropped
as USB console device is enabled in the application.

Application waits for console device until its ready by checking DTR
flag - uart_line_ctrl_get(). This function could be dropped but then
some log messages that were generated before USB device is ready
could also be dropped.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
console: usb: Remove obsolete Kconfig option.

This patch removes unused Kconfig option from console
subsystem. If application wants to wait until the console
port is connected, enabled and ready to receive data
it should use uart_line_ctrl_get() API function and
check for DTR flag.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
samples: tracing: Enable USB by the application

usb_enable() must be called by the application.
The application may want to register usb_dc_status_callback
and trace usb status codes (usb_dc_status_code).

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
tracing: Check if init funciton is present.

Tracing backed is not checking if the chosen backed specifies
the init function and is asuming its always present which could
not be true.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
samples: zperf: usb: Enable USB by the application.

This commit allows let build zperf sample with overlay-netusb.conf.
USB subsystem must be enabled by the application.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
samples: openthread: Call usb_enable() by the app.

This commit updates openthread sample to call usb_enable() from
the application.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

carlescufi commented on samples/net/zperf/src/main.c:
```diff
@@ -0,0 +1,29 @@
+
+	(void)net_config_init_app(NULL, "Initializing network");
+#endif /* CONFIG_USB */
+}
```
Newline missing

emob-nordic replied:
Added.

carlescufi commented:
I wonder if there are other places outside of `samples/` where we need to enable USB manually now. For example:
https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/arm/atmel_sam0/common/bossa.c#L46

cc @nzmichaelh

carlescufi reviewed (changes requested):
Although technically you are not changing an API, you should still add a note to the release notes that explains the change and describes how to adapt to it.

emob-nordic commented:
> Although technically you are not changing an API, you should still add a note to the release notes that explains the change and describes how to adapt to it.

I believe You are refering to f7d166bc75. Tbh I am not 100% convinced We should introduce new API for checking if console is ready, same thing could be achieved by calling `uart_line_ctrl_get(dev, UART_LINE_CTRL_DTR, &dtr);` but the user must then get binding of the device. The addition I made is addressing @pabigot suggestion [comment](https://github.com/zephyrproject-rtos/zephyr/issues/27071#issuecomment-665574602).

I let that go and I am waiting for opinions.

This addition requires users to `#include <drivers/console/uart_console.h>`.
If the app wishes to use more line controls (Kconfig option) like `UART_LINE_CTRL_BAUD_RATE` it will have to `#include <drivers/uart.h>` in addition and call `uart_line_ctrl_get(dev, control, &ret);`

Furthermore as pointed by @pabigot in #27160 it seems that not only DTR is the the thing we should wait for to say: "hey, Your console is ready now". I had just a brief look into that and  there is one suspect but didn't have time to investage any further.

Commit:
doc: releases: 2.4: Add USB detailed changes.

Add a entry to 2.4 release notes about a need to explicitly
call usb_enable() for applications that configures specific
features.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

emob-nordic commented:
> Although technically you are not changing an API, you should still add a note to the release notes that explains the change and describes how to adapt to it.

After a rethink I brought back changes to `tracing_backend.h`. I believe there is no point to add `*_is_ready` API function. The only change I made is check if `init` function exists prior to calling it and this I included in the release notes.

emob-nordic commented:
> I wonder if there are other places outside of `samples/` where we need to enable USB manually now. For example:
> https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/arm/atmel_sam0/common/bossa.c#L46
> 
> cc @nzmichaelh

I don't think I can do something about that. The Kconfig option for bossa bootloader says:
```
Signifies that the target uses a BOSSA compatible bootloader.  If CDC
ACM USB support is also enabled then the board will reboot into the
bootloader automatically when bossac is run.
```

I assume once user enabled CDC ACM alongside bossa bootloader he/she will need to enable the USB subsystem as described in API notes.

carlescufi commented:
> I believe You are refering to [f7d166b](https://github.com/zephyrproject-rtos/zephyr/commit/f7d166bc7570184976ebf28127e6d393d29b068b). Tbh I am not 100% convinced We should introduce new API for checking if console is ready, same thing could be achieved by calling `uart_line_ctrl_get(dev, UART_LINE_CTRL_DTR, &dtr);` but the user must then get binding of the device. The addition I made is addressing @pabigot suggestion [comment](https://github.com/zephyrproject-rtos/zephyr/issues/27071#issuecomment-665574602).

No no, I was referring to the fact that you now need to call `usb_enable()` manually even if using cdc acm or other classes. This should definitely be in the release notes, because it forces the user to modify their application in order to achieve the same result. We should list the cases and Kconfig options that now require enabling USB from the app, and that before did not.

pabigot commented:
> No no, I was referring to the fact that you now need to call `usb_enable()` manually even if using cdc acm or other classes. 

Users have been required to invoke `usb_enable()` explicitly since 2.2 was released.  That there were exceptions was an error.  I agree that it should be noted in the release notes that more applications will need to change (and note the resulting increased impact on lossage from USB console due to it starting later).

> I assume once user enabled CDC ACM alongside bossa bootloader he/she will need to enable the USB subsystem as described in API notes.

The callback for bossa allows USB to be used like a reset button.  AFAIK the boot loader doesn't need USB, or if it does it can enable it.  Also as long as the CDC ACM device can be obtained at the point the callback is registered it shouldn't matter whether at that point the device is ready to be used as a console.

(I don't know if the CDC ACM device pointer can be obtained before the device initializes; USB and normal devices may be different that way.)

I haven't had a chance to look at the implementation here; I will try to do so early this week.

jfischer-phytec-iot commented on samples/boards/sensortile_box/src/main.c:
```diff
@@ -248,8 +250,22 @@ static void iis3dhhc_config(struct device *iis3dhhc)
+	}
+
+	/* Poll if the DTR flag was set, optional */
+	while (!dtr) {
```
Is variable dtr used uninitialized?

emob-nordic replied:
whoops.. corrected.

jfischer-phytec-iot commented on samples/net/zperf/src/main.c:
```diff
@@ -0,0 +1,29 @@
+ * @brief Zperf sample.
+ */
+
+#if defined(CONFIG_USB)
```
Do we really need to guard this two includes?

emob-nordic replied:
No. I've removed all conditional includes.

pabigot reviewed (changes requested):
I do think we need a `cdc_acm_uart_ready()` API function and that the examples should use it.  The console in particular will not work well if material gets output before it's ready.

IMO it shouldn't be `uart_console_ready()` because it's not generic to uart, and it's not specific to console.  It is generic to CDC_ACM.

emob-nordic commented on samples/subsys/tracing/src/main.c:
```diff
@@ -7,6 +7,10 @@
 #include <zephyr.h>
 #include <sys/printk.h>
+#include <logging/log.h>
+#if (defined(CONFIG_USB) && defined(CONFIG_USB_DEVICE_STACK))
```
> I do think we need a `cdc_acm_uart_ready()` API function and that the examples should use it. The console in particular will not work well if material gets output before it's ready.
> 
> IMO it shouldn't be `uart_console_ready()` because it's not generic to uart, and it's not specific to console. It is generic to CDC_ACM.

By this patch I don't want to address the #27160 but just make `usb_enable()` be called across all the samples that use USB.

All other comments are addressed.

emob-nordic replied:
I've removed all conditional includes.

emob-nordic commented:
> I do think we need a `cdc_acm_uart_ready()` API function and that the examples should use it. The console in particular will not work well if material gets output before it's ready.
> 
> IMO it shouldn't be `uart_console_ready()` because it's not generic to uart, and it's not specific to console. It is generic to CDC_ACM.

By this patch I don't want to address the #27160 but just make `usb_enable()` be called across all the samples that use USB.

All other comments are addressed.

emob-nordic commented on samples/boards/sensortile_box/src/main.c:
```diff
@@ -248,8 +250,22 @@ static void iis3dhhc_config(struct device *iis3dhhc)
+			CONFIG_UART_CONSOLE_ON_DEV_NAME);
 	int i, on = 1;
 	int cnt = 1;
+	uint32_t dtr;
```
done

emob-nordic commented on samples/boards/sensortile_box/src/main.c:
```diff
@@ -248,8 +250,22 @@ static void iis3dhhc_config(struct device *iis3dhhc)
+	/* Poll if the DTR flag was set, optional */
+	while (!dtr) {
+		uart_line_ctrl_get(dev, UART_LINE_CTRL_DTR, &dtr);
+		k_msleep(100);
```
done

jfischer-phytec-iot replied:
k_msleep(100); here is to relax situation and give cpu time for the logging thread.

emob-nordic replied:
Yes, but this DTR is waiting for console to actually log those so If we put this k_msleep() nothing will be send anyway, because console is not yet active.

emob-nordic commented on drivers/console/Kconfig:
```diff
@@ -92,14 +92,6 @@ config USB_UART_CONSOLE
 	  layer currently doesn't support multiple interfaces, this shouldn't be
 	  selected in conjunction with, say, USB Mass Storage.
```
Sorry, it's a relic I forgot to delete - updated.

emob-nordic commented:
> Although technically you are not changing an API, you should still add a note to the release notes that explains the change and describes how to adapt to it.

I've made a note to `release-notes-2.4.rst` file.

pabigot reviewed (approved):
Seems fine.

I have requirements on the TBD function used to wait for the CDC ACM UART to be ready, and on the role of the callback passed to `usb_enable()`.  I'll open other issues for those.

jfischer-phytec-iot commented on subsys/net/lib/config/Kconfig:
```diff
@@ -5,6 +5,7 @@
 
 config NET_CONFIG_AUTO_INIT
 	bool "Init networking support automatically during device startup"
+	default n if USB_DEVICE_NETWORK_ECM
```
I thing it should be
```
default n if USB_DEVICE_NETWORK
```
because netusb handles all tree types, ECM, EEM, and RNDIS

emob-nordic replied:
done.
